### ISSUE TITLE
Conn and Stream should provide Contexts

### ISSUE THREAD

**Issue by lthibault** [libp2p/go-libp2p-net#32]
It's often useful to have some idea of when a connection is lost.  This is especially true when working with multiplexed connections where a given stream may become unusable because it's underlying connection has been closed.

I'm quite fond of the approach taken by https://github.com/lucas-clemente/quic-go.  QUIC-go's [Session](https://godoc.org/github.com/lucas-clemente/quic-go#Session) and [Stream](https://godoc.org/github.com/lucas-clemente/quic-go#Stream) interfaces export the `Context()` method.  The returned context expires when the Session/Stream is no longer usable.

Does this seem feasible in libp2p?

**Comment by Stebalien** [comment]
Yeah, this would be especially useful for aborting requests. However, we have to be careful about *when* we cancel the context. Really, I think we'd only want to cancel it when the stream is reset (not just closed).

We'd also have to carefully think about performance.

TL;DR: I agree this would be useful, it will be a pretty significant addition to the API, and we have to think about it carefully.

**Comment by lthibault** [comment]
> Really, I think we'd only want to cancel it when the stream is reset (not just closed).


Reset = "fully closed" & closed = "half-closed", correct?  (As per the Stream interface?)

If so, then I fully agree.

> I agree this would be useful, it will be a pretty significant addition to the API, and we have to think about it carefully.

What are the next steps? Anything I can do on my end?

**Comment by Stebalien** [comment]
> Reset = "fully closed" & closed = "half-closed", correct? (As per the Stream interface?)

Almost. Full closed is: I closed and then *observed* your close. Reset is: I called reset (or you called reset).

> What are the next steps? Anything I can do on my end?

1. Try to get a consensus (give motivations/use-cases).
2. Propose a concrete interface change.
3. Get a consensus on that.
4. Make the changes to the underlying transports.
5. Get those merged.
6. Finally, change the interface.

Basically, I *think* this is a step in the right direction and would kind of like to see it happen happen but I don't really have the time to push for it at the moment.

@libp2p/go-team ?

**Comment by hsanjuan** [comment]
You can get the idea that a connection is lost by reading or writing from it (and handling the associated error). It is a nice to have but I don't think having this unlocks doing something that is fully impossible right now? (do correct me here if I'm wrong).

What I'm saying is, this is nice, but if you're blocked, we could probably help you work around it while the feature lands.
